src/Core/Environment.php

```diff
@@ -43,7 +43,12 @@ class Environment
     public static function getVariables()
     {
         // Suppress return by-ref
-        return array_merge($GLOBALS, [ 'env' => static::$env ]);
+        $vars = [ 'env' => static::$env ];
+        foreach ($GLOBALS as $varName => $varValue) {
+            $vars[$varName] = $varValue;
+        }
+
+        return $vars;
```

Comment:
could achieve the same with `array_merge([], $GLOBALS, [ 'env' => static::$env ]);`, only the first array param is mutated.

Reply:
Try it. It doesn't work. This is either the most annoying bug of all time, or I'm missing something super obvious. Nothing makes sense.

Reply:
seems to work for me:
```
        echo '<pre>'. print_r($GLOBALS, true) .'</pre>';
        echo '<br />------------<br />';
        echo '<pre>'. print_r(array_merge([], $GLOBALS, [ 'env' => static::$env ]), true) .'</pre>';
        die();
```
results:
<details>
```
Array
(
    [_GET] => Array
        (
        )

    [_POST] => Array
        (
        )

    [_COOKIE] => Array
        (
            [PHPSESSID] => 8d2j2t9qmnehhnlfsvdein4hr7
            [alc_enc] => 93:cce003dae6c108382e30704e6c7e5af5446df432
            [alc_device] => 15e81b8a175a32682799e48513d122a411fad361
            [bypassStaticCache] => 1
        )

    [_FILES] => Array
        (
        )

    [GLOBALS] => Array
 *RECURSION*
    [__composer_autoload_files] => Array
        (
            [0e6d7bf4a5811bfa5cf40c5ccd6fae6a] => 1
            [32dcc8afd4335739640db7d200c1971d] => 1
            [2c102faa651ef8ea5874edb585946bce] => 1
            [5255c38a0faeba867671b61dfda6d864] => 1
            [795b01cf23ed94cdbfa9afcee6217952] => 1
            [a0edc8309cc5e1d60e3047b5df6b7052] => 1
            [c594688b3441835d5575f3085da4a242] => 1
            [25072dd6e2470089de65ae7bf11d3109] => 1
            [c964ee0ededf28c96ebd9db5099ef910] => 1
            [667aeda72477189d0494fecd327c3641] => 1
            [37a3dc5111fe8f707ab4c132ef1dbc62] => 1
            [6124b4c8570aa390c21fafd04a26c69f] => 1
        )

    [_SERVER] => Array
        (
            [REDIRECT_HTTP_AUTHORIZATION] => 
            [REDIRECT_HTTP_MOD_REWRITE] => On
            [REDIRECT_STATUS] => 200
            [HTTP_AUTHORIZATION] => 
            [HTTP_MOD_REWRITE] => On
            [HTTP_HOST] => ss4.test
            [HTTP_CONNECTION] => keep-alive
            [HTTP_CACHE_CONTROL] => max-age=0
            [HTTP_UPGRADE_INSECURE_REQUESTS] => 1
            [HTTP_USER_AGENT] => Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.167 Safari/537.36
            [HTTP_ACCEPT] => text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
            [HTTP_ACCEPT_ENCODING] => gzip, deflate
            [HTTP_ACCEPT_LANGUAGE] => en-US,en;q=0.9
            [HTTP_COOKIE] => PHPSESSID=8d2j2t9qmnehhnlfsvdein4hr7; alc_enc=93%3Acce003dae6c108382e30704e6c7e5af5446df432; alc_device=15e81b8a175a32682799e48513d122a411fad361; bypassStaticCache=1
            [PATH] => /usr/bin:/bin:/usr/sbin:/sbin
            [SERVER_SIGNATURE] => 
            [SERVER_SOFTWARE] => Apache/2.4.27 (Unix) PHP/7.1.8
            [SERVER_NAME] => ss4.test
            [SERVER_ADDR] => 127.0.0.1
            [SERVER_PORT] => 80
            [REMOTE_ADDR] => 127.0.0.1
            [DOCUMENT_ROOT] => /Users/cjoe/Sites/silverstripe/ss4
            [REQUEST_SCHEME] => http
            [CONTEXT_PREFIX] => 
            [CONTEXT_DOCUMENT_ROOT] => /Users/cjoe/Sites/silverstripe/ss4
            [SERVER_ADMIN] => you@example.com
            [SCRIPT_FILENAME] => /Users/cjoe/Sites/silverstripe/ss4/index.php
            [REMOTE_PORT] => 51655
            [REDIRECT_URL] => /admin/pages/
            [GATEWAY_INTERFACE] => CGI/1.1
            [SERVER_PROTOCOL] => HTTP/1.1
            [REQUEST_METHOD] => GET
            [QUERY_STRING] => 
            [REQUEST_URI] => /admin/pages/
            [SCRIPT_NAME] => /index.php
            [PHP_SELF] => /index.php
            [REQUEST_TIME_FLOAT] => 1519597290.375
            [REQUEST_TIME] => 1519597290
        )

    [_ENV] => Array
        (
        )

)

------------
Array
(
    [_GET] => Array
        (
        )

    [_POST] => Array
        (
        )

    [_COOKIE] => Array
        (
            [PHPSESSID] => 8d2j2t9qmnehhnlfsvdein4hr7
            [alc_enc] => 93:cce003dae6c108382e30704e6c7e5af5446df432
            [alc_device] => 15e81b8a175a32682799e48513d122a411fad361
            [bypassStaticCache] => 1
        )

    [_FILES] => Array
        (
        )

    [GLOBALS] => Array
        (
            [_GET] => Array
                (
                )

            [_POST] => Array
                (
                )

            [_COOKIE] => Array
                (
                    [PHPSESSID] => 8d2j2t9qmnehhnlfsvdein4hr7
                    [alc_enc] => 93:cce003dae6c108382e30704e6c7e5af5446df432
                    [alc_device] => 15e81b8a175a32682799e48513d122a411fad361
                    [bypassStaticCache] => 1
                )

            [_FILES] => Array
                (
                )

            [GLOBALS] => Array
 *RECURSION*
            [__composer_autoload_files] => Array
                (
                    [0e6d7bf4a5811bfa5cf40c5ccd6fae6a] => 1
                    [32dcc8afd4335739640db7d200c1971d] => 1
                    [2c102faa651ef8ea5874edb585946bce] => 1
                    [5255c38a0faeba867671b61dfda6d864] => 1
                    [795b01cf23ed94cdbfa9afcee6217952] => 1
                    [a0edc8309cc5e1d60e3047b5df6b7052] => 1
                    [c594688b3441835d5575f3085da4a242] => 1
                    [25072dd6e2470089de65ae7bf11d3109] => 1
                    [c964ee0ededf28c96ebd9db5099ef910] => 1
                    [667aeda72477189d0494fecd327c3641] => 1
                    [37a3dc5111fe8f707ab4c132ef1dbc62] => 1
                    [6124b4c8570aa390c21fafd04a26c69f] => 1
                )

            [_SERVER] => Array
                (
                    [REDIRECT_HTTP_AUTHORIZATION] => 
                    [REDIRECT_HTTP_MOD_REWRITE] => On
                    [REDIRECT_STATUS] => 200
                    [HTTP_AUTHORIZATION] => 
                    [HTTP_MOD_REWRITE] => On
                    [HTTP_HOST] => ss4.test
                    [HTTP_CONNECTION] => keep-alive
                    [HTTP_CACHE_CONTROL] => max-age=0
                    [HTTP_UPGRADE_INSECURE_REQUESTS] => 1
                    [HTTP_USER_AGENT] => Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.167 Safari/537.36
                    [HTTP_ACCEPT] => text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
                    [HTTP_ACCEPT_ENCODING] => gzip, deflate
                    [HTTP_ACCEPT_LANGUAGE] => en-US,en;q=0.9
                    [HTTP_COOKIE] => PHPSESSID=8d2j2t9qmnehhnlfsvdein4hr7; alc_enc=93%3Acce003dae6c108382e30704e6c7e5af5446df432; alc_device=15e81b8a175a32682799e48513d122a411fad361; bypassStaticCache=1
                    [PATH] => /usr/bin:/bin:/usr/sbin:/sbin
                    [SERVER_SIGNATURE] => 
                    [SERVER_SOFTWARE] => Apache/2.4.27 (Unix) PHP/7.1.8
                    [SERVER_NAME] => ss4.test
                    [SERVER_ADDR] => 127.0.0.1
                    [SERVER_PORT] => 80
                    [REMOTE_ADDR] => 127.0.0.1
                    [DOCUMENT_ROOT] => /Users/cjoe/Sites/silverstripe/ss4
                    [REQUEST_SCHEME] => http
                    [CONTEXT_PREFIX] => 
                    [CONTEXT_DOCUMENT_ROOT] => /Users/cjoe/Sites/silverstripe/ss4
                    [SERVER_ADMIN] => you@example.com
                    [SCRIPT_FILENAME] => /Users/cjoe/Sites/silverstripe/ss4/index.php
                    [REMOTE_PORT] => 51655
                    [REDIRECT_URL] => /admin/pages/
                    [GATEWAY_INTERFACE] => CGI/1.1
                    [SERVER_PROTOCOL] => HTTP/1.1
                    [REQUEST_METHOD] => GET
                    [QUERY_STRING] => 
                    [REQUEST_URI] => /admin/pages/
                    [SCRIPT_NAME] => /index.php
                    [PHP_SELF] => /index.php
                    [REQUEST_TIME_FLOAT] => 1519597290.375
                    [REQUEST_TIME] => 1519597290
                )

            [_ENV] => Array
                (
                )

        )

    [__composer_autoload_files] => Array
        (
            [0e6d7bf4a5811bfa5cf40c5ccd6fae6a] => 1
            [32dcc8afd4335739640db7d200c1971d] => 1
            [2c102faa651ef8ea5874edb585946bce] => 1
            [5255c38a0faeba867671b61dfda6d864] => 1
            [795b01cf23ed94cdbfa9afcee6217952] => 1
            [a0edc8309cc5e1d60e3047b5df6b7052] => 1
            [c594688b3441835d5575f3085da4a242] => 1
            [25072dd6e2470089de65ae7bf11d3109] => 1
            [c964ee0ededf28c96ebd9db5099ef910] => 1
            [667aeda72477189d0494fecd327c3641] => 1
            [37a3dc5111fe8f707ab4c132ef1dbc62] => 1
            [6124b4c8570aa390c21fafd04a26c69f] => 1
        )

    [_SERVER] => Array
        (
            [REDIRECT_HTTP_AUTHORIZATION] => 
            [REDIRECT_HTTP_MOD_REWRITE] => On
            [REDIRECT_STATUS] => 200
            [HTTP_AUTHORIZATION] => 
            [HTTP_MOD_REWRITE] => On
            [HTTP_HOST] => ss4.test
            [HTTP_CONNECTION] => keep-alive
            [HTTP_CACHE_CONTROL] => max-age=0
            [HTTP_UPGRADE_INSECURE_REQUESTS] => 1
            [HTTP_USER_AGENT] => Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.167 Safari/537.36
            [HTTP_ACCEPT] => text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
            [HTTP_ACCEPT_ENCODING] => gzip, deflate
            [HTTP_ACCEPT_LANGUAGE] => en-US,en;q=0.9
            [HTTP_COOKIE] => PHPSESSID=8d2j2t9qmnehhnlfsvdein4hr7; alc_enc=93%3Acce003dae6c108382e30704e6c7e5af5446df432; alc_device=15e81b8a175a32682799e48513d122a411fad361; bypassStaticCache=1
            [PATH] => /usr/bin:/bin:/usr/sbin:/sbin
            [SERVER_SIGNATURE] => 
            [SERVER_SOFTWARE] => Apache/2.4.27 (Unix) PHP/7.1.8
            [SERVER_NAME] => ss4.test
            [SERVER_ADDR] => 127.0.0.1
            [SERVER_PORT] => 80
            [REMOTE_ADDR] => 127.0.0.1
            [DOCUMENT_ROOT] => /Users/cjoe/Sites/silverstripe/ss4
            [REQUEST_SCHEME] => http
            [CONTEXT_PREFIX] => 
            [CONTEXT_DOCUMENT_ROOT] => /Users/cjoe/Sites/silverstripe/ss4
            [SERVER_ADMIN] => you@example.com
            [SCRIPT_FILENAME] => /Users/cjoe/Sites/silverstripe/ss4/index.php
            [REMOTE_PORT] => 51655
            [REDIRECT_URL] => /admin/pages/
            [GATEWAY_INTERFACE] => CGI/1.1
            [SERVER_PROTOCOL] => HTTP/1.1
            [REQUEST_METHOD] => GET
            [QUERY_STRING] => 
            [REQUEST_URI] => /admin/pages/
            [SCRIPT_NAME] => /index.php
            [PHP_SELF] => /index.php
            [REQUEST_TIME_FLOAT] => 1519597290.375
            [REQUEST_TIME] => 1519597290
        )

    [_ENV] => Array
        (
        )

    [env] => Array
        (
            [SS_DATABASE_CLASS] => MySQLDatabase
            [SS_DATABASE_SERVER] => 127.0.0.1
            [SS_DATABASE_USERNAME] => root
            [SS_DATABASE_PASSWORD] => rootpass
            [SS_DATABASE_PREFIX] => SS_silverstripe
            [SS_DATABASE_NAME] => 4
            [SS_ENVIRONMENT_TYPE] => dev
            [SS_DEFAULT_ADMIN_USERNAME] => admin
            [SS_DEFAULT_ADMIN_PASSWORD] => password
            [SS_BASE_URL] => http://localhost:8080/
            [SS_VENDOR_METHOD] => symlink
        )

)
```
</details>

Reply:
Right, but you need to test it against the rewrite of the global session:

```
// Setup session
        if ($session instanceof Session) {
                    // ........
        } else {
           var_dump($_SESSION) // populated
            $newVars['_SESSION'] = $session ?: [];
            var_dump($_SESSION); // empty

        }

```

Reply:
Oh, I see...
Your test still passes if I switched to `array_merge([], $GLOBALS)`... I'd accept if there was a comment above the loop to express that `array_merge()` isn't reliable.

Not sure how easy it would be have a reproducible test scenario for a ~super~global... so a comment will suffix.

Reply:
Yeah, I just realised that, too. The test is useless. I'm not sure it's worth investigating a way to isolate that error. All I really know is that the error goes away when using that change to `getVariables()` at zero cost, from what I can tell. But yeah, this one made my head spin.

Reply:
Your test needs to modify an actual global. E.g. `_SESSION` not a dummy var.

Reply:
Actually I can't reproduce the bug at all. Wonder why...